File: src/packages/skill-hermes/agentxp_skill_hermes/identity.py
"""On-disk identity material.

Matches `src/packages/skill/src/identity.ts` byte-for-byte at the file
layer: operator.json / agent.json with hex-encoded 32-byte seeds.
SPEC 03-modules-product §4 acceptance 2: running Skill-Hermes against
the same identity directory as Skill MUST produce events attributable
to the same operator_pubkey.
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from .keys import AgentKey, OperatorKey, delegate_agent_key, generate_operator_key


class OperatorKeyMissingError(FileNotFoundError):
    pass


@dataclass(frozen=True)
class IdentityPaths:
    root: Path
    operator_file: Path
    agent_file: Path


def _default_root() -> Path:
    return Path.home() / ".agentxp" / "identity"


def resolve_identity_paths(root: Optional[Path | str] = None) -> IdentityPaths:
    r = Path(root) if root is not None else _default_root()
    return IdentityPaths(
        root=r,
        operator_file=r / "operator.json",
        agent_file=r / "agent.json",
    )


def _write_private(path: Path, data: dict) -> None:
    path.write_text(json.dumps(data, indent=2))
    try:
        os.chmod(path, 0o600)
    except OSError:
        pass


def ensure_operator_key(root: Optional[Path | str] = None) -> OperatorKey:
    paths = resolve_identity_paths(root)
    paths.root.mkdir(parents=True, exist_ok=True, mode=0o700)
    if paths.operator_file.exists():
        raw = json.loads(paths.operator_file.read_text())
        return OperatorKey(public_key=raw["publicKey"], private_key=bytes.fromhex(raw["privateKey"]))
    op = generate_operator_key()
    _write_private(
        paths.operator_file,
        {
            "publicKey": op.public_key,
            "privateKey": op.private_key.hex(),
            "created_at": int(time.time()),
        },
    )
    return op


def load_operator_key(root: Optional[Path | str] = None) -> OperatorKey:
    paths = resolve_identity_paths(root)
    if not paths.operator_file.exists():
        raise OperatorKeyMissingError(f"operator key not found at {paths.operator_file}")
    raw = json.loads(paths.operator_file.read_text())
    return OperatorKey(public_key=raw["publicKey"], private_key=bytes.fromhex(raw["privateKey"]))


def ensure_agent_key(
    operator: OperatorKey,
    agent_id: str,
    ttl_days: int,
    root: Optional[Path | str] = None,
) -> AgentKey:
    paths = resolve_identity_paths(root)
    paths.root.mkdir(parents=True, exist_ok=True, mode=0o700)
    if paths.agent_file.exists():
        raw = json.loads(paths.agent_file.read_text())
        # Mirror the TS 1-hour buffer before rolling the agent key.
        if raw["expiresAt"] > int(time.time()) + 3600:
            return AgentKey(
                public_key=raw["publicKey"],
                private_key=bytes.fromhex(raw["privateKey"]),
                delegated_by=raw["delegatedBy"],
                expires_at=raw["expiresAt"],
                agent_id=raw.get("agentId", agent_id),
            )
    agent = delegate_agent_key(operator, agent_id, ttl_days)
    _write_private(
        paths.agent_file,
        {
            "publicKey": agent.public_key,
            "privateKey": agent.private_key.hex(),
            "delegatedBy": agent.delegated_by,
            "expiresAt": agent.expires_at,
            "agentId": agent.agent_id or agent_id,
        },
    )
    return agent

File: src/packages/skill/src/identity.ts
// Local identity material for the Skill CLI.
// Operator key is the human-held master; an agent key is delegated on
// first use. SPEC 03-modules-product §3: private key bytes MUST NOT
// appear in any outgoing request; they live under ~/.agentxp/identity/.
import { mkdirSync, readFileSync, writeFileSync, existsSync, chmodSync } from 'node:fs'
import { homedir } from 'node:os'
import { join } from 'node:path'
import {
  delegateAgentKey,
  generateOperatorKey,
  bytesToHex,
  hexToBytes,
} from '@agentxp/protocol'
import type { AgentKey, OperatorKey } from '@agentxp/protocol'

function defaultRoot(): string {
  return join(homedir(), '.agentxp', 'identity')
}

export class OperatorKeyMissingError extends Error {
  constructor(path: string) {
    super(`operator key not found at ${path}`)
    this.name = 'OperatorKeyMissingError'
  }
}

export interface IdentityPaths {
  root: string
  operatorFile: string
  agentFile: string
}

export function resolveIdentityPaths(root?: string): IdentityPaths {
  const r = root ?? defaultRoot()
  return {
    root: r,
    operatorFile: join(r, 'operator.json'),
    agentFile: join(r, 'agent.json'),
  }
}

interface OperatorOnDisk {
  publicKey: string
  privateKey: string
  created_at: number
}

interface AgentOnDisk {
  publicKey: string
  privateKey: string
  delegatedBy: string
  expiresAt: number
  agentId: string
}

function writePrivate(path: string, data: unknown): void {
  writeFileSync(path, JSON.stringify(data, null, 2), { mode: 0o600 })
  try {
    chmodSync(path, 0o600)
  } catch {
    // best-effort on platforms without chmod
  }
}

export async function ensureOperatorKey(root?: string): Promise<OperatorKey> {
  const paths = resolveIdentityPaths(root)
  mkdirSync(paths.root, { recursive: true, mode: 0o700 })
  if (existsSync(paths.operatorFile)) {
    const raw = JSON.parse(readFileSync(paths.operatorFile, 'utf8')) as OperatorOnDisk
    return { publicKey: raw.publicKey, privateKey: hexToBytes(raw.privateKey) }
  }
  const op = await generateOperatorKey()
  const record: OperatorOnDisk = {
    publicKey: op.publicKey,
    privateKey: bytesToHex(op.privateKey),
    created_at: Math.floor(Date.now() / 1000),
  }
  writePrivate(paths.operatorFile, record)
  return op
}

export function loadOperatorKey(root?: string): OperatorKey {
  const paths = resolveIdentityPaths(root)
  if (!existsSync(paths.operatorFile)) {
    throw new OperatorKeyMissingError(paths.operatorFile)
  }
  const raw = JSON.parse(readFileSync(paths.operatorFile, 'utf8')) as OperatorOnDisk
  return { publicKey: raw.publicKey, privateKey: hexToBytes(raw.privateKey) }
}

export async function ensureAgentKey(
  operator: OperatorKey,
  agentId: string,
  ttlDays: number,
  root?: string,
): Promise<AgentKey> {
  const paths = resolveIdentityPaths(root)
  mkdirSync(paths.root, { recursive: true, mode: 0o700 })
  if (existsSync(paths.agentFile)) {
    const raw = JSON.parse(readFileSync(paths.agentFile, 'utf8')) as AgentOnDisk
    if (raw.expiresAt > Math.floor(Date.now() / 1000) + 3600) {
      return {
        publicKey: raw.publicKey,
        privateKey: hexToBytes(raw.privateKey),
        delegatedBy: raw.delegatedBy,
        expiresAt: raw.expiresAt,
        agentId: raw.agentId,
      }
    }
  }
  const agent = await delegateAgentKey(operator, agentId, ttlDays)
  const record: AgentOnDisk = {
    publicKey: agent.publicKey,
    privateKey: bytesToHex(agent.privateKey),
    delegatedBy: agent.delegatedBy,
    expiresAt: agent.expiresAt,
    agentId: agent.agentId ?? agentId,
  }
  writePrivate(paths.agentFile, record)
  return { ...agent, agentId: agent.agentId ?? agentId }
}
